SwiftCompilerSources/Sources/Optimizer/FunctionPasses/DiagnoseLifetimeDependence.swift

```diff
@@ -0,0 +1,570 @@
+///
+/// Lifetime *propagation* occurs when the parent lifetime is itself nonescapable. Finding the non-escaping scope for lifetime-propagation involves recursively asking for the parent's introducer. In this example, recursion gives us %outer as the next introducer, and %access as the next parent lifetime:
+///
+///     %outer = apply @returnsNonEscapable(%access)
+///               : $(@resultDependsOn @in Container) -> NonEscapableStruct
+///
+/// Lifetime *encapsulation* (as opposed to propagation) occurs when the parent lifetime is escapable. In this case, we find the parent's enclosing scope as follows:
+///
+/// For owned values, the value's lifetime is itself the enclosing scope, delimited by the consumes after following any forwarding operations.
+///
+/// For borrowed values, the enclosing scope is created by the borrow introducer (visitBorrowIntroducers).
+///
+/// For addresses, the enclosing scope is created by an access scope (enclosingAcessScope). Note that an exclusive access scope *must* exist for all non-escaping values that depend on an addressible value.
+///
+/// In this example, the parent value is an address, so the non-escaping scope is an access scope:
+///
+///     %access = begin_access %address : $*Container
+///
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+// DiagnosticEngine already declares DiagID. Why isn't it visible?
+import ASTBridging
+public typealias DiagID = BridgedDiagID
+
+private let verbose = true
+
+private func log(_ message: @autoclosure () -> String) {
+  if verbose {
+    print("### \(message())")
+  }
+}
+
+extension InstructionRange {
+  // TODO: InstructionRange strangely does not support phis or function args. This is problematic, because the block may be mutated.
+  static func getRepresentativeInstruction(value: Value) -> Instruction {
+    if let def = value.definingInstruction {
+      return def
+    }
+    if let result = TermResult(value) {
+      return result.terminator
+    }
+    assert(Phi(value) != nil || value is FunctionArgument)
+    return value.parentBlock.instructions.first!
+  }
+}
+
+/// Diagnostic pass.
+///
+/// Find the roots of all non-escapable values in this function. All non-escapable values either depend on a NonEscapingScope, or they are produced by a LifetimeDependentInstruction that has no dependence on a parent value (@_unsafeNonEscapableResult).
+let diagnoseLifetimeDependencePass = FunctionPass(
+  name: "diagnose-lifetime-dependence")
+{ (function: Function, context: FunctionPassContext) in
+  log("Diagnosing lifetime dependence in \(function.name)")
+  
+  var diagnostic = DiagnoseLifetimeDependence(context)
+  defer { diagnostic.deinitialize() }
+  // Finding the scopes of all non-escapable values would be overkill. They are all rooted in either a non-escapable function argument or a LifetimeDependentInstruction, so use those as starting points.
```

Comment:
Why? What about allocations, for example? Or do you assume that when this pass runs allocations are within (not inlined) initializer functions?

Reply:
I should get to allocations soon. This does not handle addressible dependent values at all (only addressible parents). Initially I was only trying to model lifetime-dependence between a value and some parent object. I think a SILGen post-pass should insert a new `lifetime_dependence` instruction and make sure that all `~Escapable` types are covered by at least one `lifetime_dependence`. If a value is not escapable, but has no dependence, then it should be marked dependent on itself (a lifetime-dependent orphan). SIL passes generally shouldn't need to worry about the AST type. This diagnostic would only look for those instructions.